tests/Bedrock.Framework.Tests/MessagePipeReaderTests.cs

```diff
@@ -0,0 +1,1101 @@
+
+        [Fact]
+        public async Task CanConsumeNoBytes()
+        {
+            var reader = await CreateReaderOverBytes(new byte[100]).ConfigureAwait(false);
+            var buffer = (await reader.ReadAsync()).Buffer;
+
+            reader.AdvanceTo(buffer.Start);
+
+            reader.Complete();
+        }
+
+        [Fact]
+        public async Task CanExamineAllBytes()
+        {
+            var reader = await CreateReaderOverBytes(new byte[100]).ConfigureAwait(false);
+            var buffer = (await reader.ReadAsync()).Buffer;
+
+            reader.AdvanceTo(buffer.Start, buffer.End);
+
+            reader.Complete();
+        }
+
+        [Fact]
+        public async Task CanExamineNoBytes()
+        {
+            var reader = await CreateReaderOverBytes(new byte[100]).ConfigureAwait(false);
+            var buffer = (await reader.ReadAsync()).Buffer;
+
+            reader.AdvanceTo(buffer.Start, buffer.Start);
+
+            reader.Complete();
+        }
+
+        [Fact]
+        public async Task ReadAsyncAfterReceivingCompletedReadResultDoesNotThrow()
+        {
+            var stream = new ThrowAfterZeroByteReadStream();
+            var reader = new MessagePipeReader(PipeReader.Create(stream), new TestProtocol());
+            var readResult = await reader.ReadAsync();
+            Assert.True(readResult.Buffer.IsEmpty);
+            Assert.True(readResult.IsCompleted);
+            reader.AdvanceTo(readResult.Buffer.End);
+
+            readResult = await reader.ReadAsync();
+            Assert.True(readResult.Buffer.IsEmpty);
+            Assert.True(readResult.IsCompleted);
+            reader.AdvanceTo(readResult.Buffer.End);
+            reader.Complete();
+        }
+
+        [Fact]
+        public async Task BufferingDataPastEndOfStreamCanBeReadAgain()
+        {
+            var protocol = new TestProtocol();
+            var stream = new ThrowAfterZeroByteReadStream();
+            var writer = PipeWriter.Create(stream);
+            var reader = new MessagePipeReader(PipeReader.Create(stream), protocol);
+
+            protocol.WriteMessage(Encoding.ASCII.GetBytes("Hello World"), writer);
+            await writer.FlushAsync().ConfigureAwait(false);
+            stream.Position = 0;
+
+            var readResult = await reader.ReadAsync();
+            var buffer = readResult.Buffer;
+            reader.AdvanceTo(buffer.Start, buffer.End);
+
+            // Make sure IsCompleted is true
+            readResult = await reader.ReadAsync();
+            buffer = readResult.Buffer;
+            reader.AdvanceTo(buffer.Start, buffer.End);
+            Assert.True(readResult.IsCompleted);
+
+            var value = await ReadFromPipeAsString(reader);
+            Assert.Equal("Hello World", value);
+            reader.Complete();
+        }
+
+        [Fact]
+        public async Task NextReadAfterPartiallyExaminedReturnsImmediately()
+        {
+            // TODO: decide on behaviour here, and write tests
```

Comment:
What's the deal here?

Reply:
It was in the .Net Core tests. It represents a potential optimization we could make. I'll delete it for know, and worry about optimizations later.